gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -0,0 +1,199 @@
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final StorageResourceId resourceId;
+  private WriteChannel writeChannel;
+  private boolean uploadSucceeded = false;
+  // TODO: not supported as of now
+  // private final String requesterPaysProject;
+
+  public GoogleCloudStorageClientWriteChannel(
+      Storage storage,
+      GoogleCloudStorageOptions storageOptions,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      ExecutorService uploadThreadPool) {
+    super(uploadThreadPool, storageOptions.getWriteChannelOptions());
+    this.resourceId = resourceId;
+    this.writeChannel = getClientWriteChannel(storage, resourceId, createOptions, storageOptions);
+  }
+
+  @Override
+  public void startUpload(InputStream pipeSource) throws IOException {
+    // Given that the two ends of the pipe must operate asynchronous relative
+    // to each other, we need to start the upload operation on a separate thread.
+    try {
+      uploadOperation = threadPool.submit(new UploadOperation(pipeSource, this.resourceId));
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Failed to start upload for '%s'", resourceId), e);
+    }
+  }
+
+  private static BlobInfo getBlobInfo(
+      StorageResourceId resourceId, CreateObjectOptions createOptions) {
+    BlobInfo blobInfo =
+        BlobInfo.newBuilder(
+                BlobId.of(
+                    resourceId.getBucketName(),
+                    resourceId.getObjectName(),
+                    resourceId.getGenerationId()))
+            .setContentType(createOptions.getContentType())
+            .setContentEncoding(createOptions.getContentEncoding())
+            .setMetadata(encodeMetadata(createOptions.getMetadata()))
+            .build();
+    return blobInfo;
+  }
+
+  private static WriteChannel getClientWriteChannel(
+      Storage storage,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      GoogleCloudStorageOptions storageOptions) {
+    AsyncWriteChannelOptions channelOptions = storageOptions.getWriteChannelOptions();
+    WriteChannel writeChannel =
+        storage.writer(
+            getBlobInfo(resourceId, createOptions),
+            generateWriteOptions(createOptions, channelOptions));
+    writeChannel.setChunkSize(channelOptions.getUploadChunkSize());
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final InputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(InputStream pipeSource, StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = pipeSource;
+    }
+
+    @Override
+    public Boolean call() throws Exception {
```

Comment:
I think it would be good to add some logs while we enter this method and return this method. Also some of the data like bytes written will be helpful while troubleshooting.

Reply:
Sure.